File: java/DoubleLinkedList.java
class DoubleLinkedList {
    static class Node {
        int value;
        Node next;
        Node prev;

        Node(int d) {
            value = d;
            next = null;
            prev = null;
        }
    }
    Node head;
    Node tail;

    public void insert_begin(int num) {
        if (head == null) {
            head = new Node(num);
            tail = head;
            return;
        }
        Node tmp = new Node(num);
        tmp.next = head;
        head.prev = tmp;
        head = tmp;
    }

    public void insert_end(int num) {
        if (tail == null) {
            tail = new Node(num);
            head = tail;
            return;
        }
        tail.next = new Node(num);
        tail.next.prev = tail;
        tail = tail.next;
    }

    public void insert_middle(int num, int pos) {
        if (pos == 0) {
            insert_begin(num);
            return;
        }
        int i = 1;
        for(Node curr = head; curr != null; curr = curr.next) {
            if (i == pos) {
                if (curr.next == null) {
                    insert_end(num);
                }
                Node tmp = new Node(num);
                tmp.next = curr.next;
                curr.next.prev = tmp;
                tmp.prev = curr;
                curr.next = tmp;
                return;
            }
            i++;
        }
    }

    public void delete_begin() {
        head = head.next;
        head.prev = null;
    }

    public void delete_end() {
        if (tail.prev == null) {
            tail = null;
            return;
        }
        tail = tail.prev;
        tail.next = null;
    }

    public void delete_middle(int pos) {
        if (pos == 0) {
            delete_begin();
            return;
        }
        int i = 1;
        for (Node curr = head; curr != null; curr = curr.next) {
            if (i == pos) {
                if (curr == tail) {
                    delete_end();
                    return;
                }
                curr.prev.next = curr.next;
                curr.next.prev = curr.prev;
            }
            i++;
        }
    }

    public void traverse() {
        for (Node curr = head; curr != null; curr = curr.next) {
            System.out.print(curr.value+"->");
        }
        System.out.println();
        for (Node curr = tail; curr != null; curr = curr.prev) {
            System.out.print(curr.value+"->");
        }
        System.out.println();
    }
    public static void main(String[] args) {
        LinkedList ll = new LinkedList();
        ll.insert_begin(7);
        ll.insert_begin(1);
        ll.insert_begin(2);
        ll.traverse();
        ll.insert_end(3);
        ll.insert_end(4);
        ll.traverse();
        ll.insert_middle(8, 0);
        ll.insert_middle(8, 4);
        ll.traverse();
        ll.delete_begin();
        ll.traverse();
        ll.delete_end();
        ll.traverse();
        ll.delete_middle(3);
        ll.traverse();
    }
}

File: c++/double_linked_list.cpp
#include <iostream>
using namespace std;

struct node {
    int data;
    node* next;
    node* prev;
};
typedef struct node node;

node* create_node(int data) {
    node* new_node = new node();
    new_node->data = data;
    new_node->next = new_node->prev = nullptr;
    return new_node;
}

void insert_begin(node** head, node** tail, int data) {
    node* new_node = create_node(data);
    if (*head == nullptr) {
        *head = *tail = new_node;
    } else {
        new_node->next = *head;
        (*head)->prev = new_node;
        *head = new_node;
    }
}

void insert_end(node** head, node** tail, int data) {
    node* new_node = create_node(data);
    if (*head == nullptr) {
        *head = *tail = new_node;
    } else {
        new_node->prev = *tail;
        (*tail)->next = new_node;
        *tail = new_node;
    }
}

void insert_middle(node** head, node** tail, int data, int pos) {
    node *new_node = create_node(data);
    int i = 1;
    for (node *curr = *head; curr != nullptr; curr = curr->next) {
        if (i == pos) {
            if (curr->next == nullptr) {
                insert_end(head, tail, data);
            } 
            new_node->next = curr->next;
            new_node->prev = curr;
            curr->next = new_node;
            new_node->next->prev = new_node;
            return;
        }
    }
}

void delete_begin (node** head) {
    *head = (*head)->next;
    (*head)->prev = nullptr;
}

void delete_end (node** tail) {
    *tail = (*tail)->prev;
    (*tail)->next = nullptr;
}

void traverse(node* head, node* tail) {
    cout<<"Forward Traversal:\n";
    for (node *tmp = head; tmp != nullptr; tmp = tmp->next) {
        cout<<tmp->data<<"->";
    }
    cout<<"\nBackward Traversal:\n";
    for (node *tmp = tail; tmp != nullptr; tmp = tmp->prev) {
        cout<<tmp->data<<"<-";
    }
    cout<<"\n";
}

int main() {
    node* head = nullptr;
    node* tail = nullptr;
    insert_begin(&head, &tail, 3);
    insert_begin(&head, &tail, 4);
    traverse(head, tail);
    insert_middle(&head, &tail, 6, 1);
    traverse(head, tail);
    delete_begin(&head);
    delete_end(&tail);
    traverse(head, tail);
    return 0;
}
